src/ecs/resources/input.rs

```diff
@@ -433,82 +384,118 @@ impl InputHandler {
+                        any_action_is_pressed_this_frame = true;
+                    }
                 }
-                if all_buttons_are_pressed && !self.button_is_pressed(*button) {
-                    all_buttons_are_pressed = false;
+                if all_actions_are_pressed {
+                    if buttons.iter().all(|&b| !self.button_is_pressed(b)) {
+                        all_actions_are_pressed = false;
+                    }
                 }
             } else {
-                bad_values.push(*action);
+                bad_values.push(action.to_string());
             }
         }
         if !bad_values.is_empty() {
             Err(bad_values)
         } else {
-            Ok(all_buttons_are_pressed && any_button_is_pressed_this_frame)
+            Ok(all_actions_are_pressed && any_action_is_pressed_this_frame)
         }
     }
 
     /// Assign an axis to an ID value
     ///
     /// This will insert a new axis if no entry for this id exists.
     /// If one does exist this will replace the axis at that id and return it.
-    pub fn insert_axis(&mut self, id: i32, axis: Axis) -> Option<Axis> {
+    pub fn insert_axis(&mut self, id: String, axis: Axis) -> Option<Axis> {
         self.axes.insert(id, axis)
     }
 
     /// Removes an axis, this will return the removed axis if successful.
-    pub fn remove_axis(&mut self, id: i32) -> Option<Axis> {
-        self.axes.remove(&id)
+    pub fn remove_axis(&mut self, id: &str) -> Option<Axis> {
+        self.axes.remove(id)
     }
 
     /// Returns a reference to an axis.
-    pub fn get_axis(&mut self, id: i32) -> Option<&Axis> {
-        self.axes.get(&id)
+    pub fn axis(&mut self, id: &str) -> Option<&Axis> {
+        self.axes.get(id)
     }
 
-    /// Assign an action to an ID value
-    ///
-    /// This will insert a new action if no entry for this id exists.
-    /// If one does exist this will replace the action at that id and return it.
-    pub fn insert_action(&mut self, id: i32, binding: Button) -> Option<Button> {
-        self.actions.insert(id, binding)
+    /// Get's a list of all axes
+    pub fn axes(&self) -> Vec<String> {
+        self.axes.keys().map(|k| k.clone()).collect::<Vec<String>>()
     }
 
-    /// Removes an action, this will return the removed action if successful.
-    pub fn remove_action(&mut self, id: i32) -> Option<Button> {
-        self.actions.remove(&id)
+    /// Add a button to an action.
+    ///
+    /// This will insert a new binding between this action and the button.
+    pub fn insert_action_binding(&mut self, id: &str, binding: Button) {
+        let mut make_new = false;
+        match self.actions.get_mut(id) {
+            Some(action_bindings) => {
+                if action_bindings.iter().all(|&b| b != binding) {
+                    action_bindings.push(binding);
+                }
+            }
+            None => {
+                make_new = true;
```

Comment:
Ahh, I see. I wonder if the upcoming non-lexical lifetimes in Rust will help with these kinds of issues.